### PR TITLE
[AIRFLOW-255] schedule_dag shouldn't return early if dagrun_timeout is given.

### PR DESCRIPTION
**Pull request by reconditesea**
Dear Airflow Maintainers,

Please accept this PR that addresses the following issues:
- https://issues.apache.org/jira/browse/AIRFLOW-255
- Testing Done:  ./run_unit_tests.sh

In https://github.com/apache/incubator-airflow/blob/master/airflow/jobs.py#L396, schedule_dag returns as long as `len(active_runs) >= dag.max_active_runs`. It should also take into account dag.dagrun_timeout

### CODE REVIEW

File: airflow/jobs.py [hunk 1]
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
                 external_trigger=False,
                 session=session
             )
-            if len(active_runs) >= dag.max_active_runs:
+            # retunr if already reach maximum active runs and no timeout setting
```
Comment (bolkedebruin) [inline airflow/jobs.py]: nit: typo

File: airflow/jobs.py [hunk 2]
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
...
+            if len(active_runs) >= dag.max_active_runs and not dag.dagrun_timeout:
                 return
+            timeout_triggered = 0
             for dr in active_runs:
                 if (
                         dr.start_date and dag.dagrun_timeout and
                         dr.start_date < datetime.now() - dag.dagrun_timeout):
                     dr.state = State.FAILED
                     dr.end_date = datetime.now()
+                    timeout_triggered += 1
```
Comment (bolkedebruin) [inline airflow/jobs.py]: please rename this variable. "failed_runs" seems better to me

File: airflow/jobs.py [hunk 3]
```diff
@@ -393,14 +393,19 @@ def schedule_dag(self, dag):
...
+            if len(active_runs) - timeout_triggered >= dag.max_active_runs:
+                return
             session.commit()
```
Comment (bolkedebruin) [inline airflow/jobs.py]: the commit needs to happen before returning, now your losing the state update.

**Comment by bolkedebruin** [comment]
@reconditesea thanks for the contribution. Please adhere to the template for PRs that has been described and make sure to update your commit message with a proper description. As the scheduler is very much the core of airflow, please also provide updated tests

File: airflow/jobs.py (hunk 2 above)
Comment (reconditesea) [inline airflow/jobs.py]: How about timedout_runs? As timed out doesn't necessarily mean failure.

**Comment by reconditesea** [comment]
@bolkedebruin Thanks for the review and  suggestion! Will update the test shortly.

File: airflow/jobs.py (hunk 2 above)
Comment (bolkedebruin) [inline airflow/jobs.py]: Also fine!

**Comment by reconditesea** [comment]
@bolkedebruin Address the comments and add the unit-test 😃

**Comment by bolkedebruin** [comment]
@reconditesea Nice looks good. But please squash your commits and I think your commit message seems a bit off (it's a link).

**Comment by reconditesea** [comment]
@bolkedebruin Sure thing, done!

File: airflow/jobs.py [hunk 4]
```diff
@@ -393,15 +393,20 @@ def schedule_dag(self, dag):
                 external_trigger=False,
                 session=session
             )
-            if len(active_runs) >= dag.max_active_runs:
+            # return if already reach maximum active runs and no timeout setting
```
Comment (bolkedebruin) [inline airflow/jobs.py]: nit: reached

File: tests/jobs.py [hunk 5]
```diff
@@ -685,6 +685,45 @@ def test_scheduler_fail_dagrun_timeout(self):
...
+        session.commit()
+        session.close()
+
+        scheduler = SchedulerJob()
+        dag.clear()
+
+        dr = scheduler.schedule_dag(dag)
+        self.assertIsNotNone(dr)
+
+        # Should not be scheduled as no dagrun_timeout specified
```
Comment (bolkedebruin) [inline tests/jobs.py]: Nit: I think this assertion is incorrect: ie. should be "Should not be scheduled as DagRun has not timedout and max_active_runs is reached". It is now somewhat confusing.

File: tests/jobs.py [hunk 6]
```diff
@@ -685,6 +685,45 @@ def test_scheduler_fail_dagrun_timeout(self):
...
+        new_dr = scheduler.schedule_dag(dag)
+        self.assertIsNone(new_dr)
+
+        # Should be scheduled as dagrun_timeout is passed
```
Comment (bolkedebruin) [inline tests/jobs.py]: nit: is = has

File: airflow/jobs.py (hunk 4 above)
Comment (bolkedebruin) [inline airflow/jobs.py]: nit: reach = reached

**Comment by bolkedebruin** [comment]
@reconditesea looks great, some final nits. Thanks for adding the tests. If you can address the final issues I'll merge.

File: tests/jobs.py (hunk 5 above)
Comment (reconditesea) [inline tests/jobs.py]: Nice catch!

**Comment by reconditesea** [comment]
@bolkedebruin Done. Thank you very much for the review!

**Comment by reconditesea** [comment]
@bolkedebruin Is this good to go now? Thanks!

**Pull request closed without merging**